src/librustc_yk_sections/emit_sir.rs

```diff
@@ -393,25 +395,41 @@ fn do_generate_sir<'a, 'tcx, 'gcx>(
         _ => None,
     };
 
-    // To satisfy the reproducible build tests, the CFG must be written out in a deterministic
-    // order, thus we sort the `DefId`s first.
-    let mut sorted_def_ids: Vec<&DefId> = def_ids.iter().collect();
-    sorted_def_ids.sort();
+    // We use a worklist because as we lower MIR bodies we will discover further DefIds (e.g. call
+    // targets) which in turn will need to be processed. However, to satisfy the reproducible build
+    // tests, we must process the DefIds in a deterministic order. To that end all newly discovered
+    // work must be sorted before it is appended into the work list.
+    let mut work: Vec<DefId> = def_ids.iter().cloned().collect();
+    work.sort();
+    let mut seen =  DefIdSet::default();
+    while !work.is_empty() {
+        let def_id = work.pop().unwrap();
+        if seen.contains(&def_id) {
+            continue;
+        }
 
-    for def_id in sorted_def_ids {
-        if tcx.is_mir_available(*def_id) {
-            let mir = tcx.optimized_mir(*def_id);
-            let mut ccx = ConvCx::new(tcx, *def_id, mir);
-            let pack = ccx.lower();
+        if tcx.is_mir_available(def_id) {
+            let mir = tcx.optimized_mir(def_id);
+            let ccx = ConvCx::new(tcx, def_id, mir);
+            let (pack, mut extra_work) = ccx.lower();
 
             if let Some(ref mut e) = enc {
                 e.serialise(ykpack::Pack::Body(pack))?;
-                e.serialise(ykpack::Pack::Debug(ykpack::SirDebug::new(
-                    ccx.lower_def_id(def_id), tcx.def_path_str(*def_id))))?;
             } else {
                 write!(textdump_file.as_ref().unwrap(), "{}", pack)?;
             }
+
+            let mut extra_work_sorted: Vec<DefId> = extra_work.drain().collect();
```

Comment:
It's fine (and would probably be clearer) to override the variable `let mut extra_work = extra_work.drain().collect::<Vec<_>>()`.